File: client/src/boundary/interactor/GameD2BConverter.test.ts
import { Board, Game } from "domain/index";
import { BoundaryBoard, BoundaryGame } from "boundary/model/index";
import { GameD2BConverter } from "./GameD2BConverter";
import { BoardD2BConverter } from "./BoardD2BConverter";
import { mock, MockProxy } from "jest-mock-extended";
import { GameState } from "utils/Constants";

describe(GameD2BConverter, () => {
  let gameConverter: GameD2BConverter;
  let boardConverter: MockProxy<BoardD2BConverter>;

  beforeEach(() => {
    boardConverter = mock<BoardD2BConverter>();
    gameConverter = new GameD2BConverter(boardConverter);
  });

  it("Converts Game model to BoundaryGame model", () => {
    const board = mock<Board>();
    const expectedBoard = mock<BoundaryBoard>();
    const inputGame: Game = new Game(
      "123",
      GameState.IN_PROGRESS,
      board,
      25,
      0
    );

    boardConverter.convert.calledWith(board).mockReturnValue(expectedBoard);

    const expectedGame: BoundaryGame = new BoundaryGame(
      "123",
      GameState.IN_PROGRESS,
      boardConverter.convert(board)
    );

    const convertedGame = gameConverter.convert(inputGame);

    expect(convertedGame.id).toEqual(expectedGame.id);
    expect(convertedGame.state).toEqual(expectedGame.state);
    expect(convertedGame.board).toEqual(expectedGame.board);
  });
});

File: server/battleships/use_cases/implementation/src/test/java/eu/optas/use_cases/implementation/GameD2BConverterTest.java
package eu.optas.use_cases.implementation;

import eu.optas.domain.Board;
import eu.optas.domain.Game;
import eu.optas.use_cases.api.BoundaryBoard;
import eu.optas.use_cases.api.BoundaryGame;
import eu.optas.utils.GameState;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class GameD2BConverterTest {

    @Test
    void convert() {
        BoardD2BConverter boardD2BConverter = mock(BoardD2BConverter.class);
        GameD2BConverter gameD2BConverter = new GameD2BConverter(boardD2BConverter);

        Board board = mock(Board.class);
        BoundaryBoard boundaryBoard = mock(BoundaryBoard.class);
        when(boardD2BConverter.convert(board)).thenReturn(boundaryBoard);

        Game inputGame = new Game(
                "123",
                GameState.IN_PROGRESS,
                board,
                25,
                0
        );
        BoundaryGame expectedGame = new BoundaryGame(
                "123",
                GameState.IN_PROGRESS,
                boundaryBoard,
                25,
                0
        );
        BoundaryGame convertedGame = gameD2BConverter.convert(inputGame);

        assertThat(convertedGame).usingRecursiveComparison().isEqualTo(expectedGame);
    }
}